src/sentry/spans/buffer.py

```diff
@@ -310,6 +320,37 @@ def _group_by_parent(self, spans: Sequence[Span]) -> dict[tuple[str, str], list[
 
         return trees
 
+    def _compress_span_payloads(self, payloads: list[bytes]) -> bytes:
+        combined = b"\x00".join(payloads)
+
+        # If compression is disabled, return the combined data as-is
+        if self._zstd_compressor is None:
+            return combined
+
+        original_size = len(combined)
+
+        with metrics.timer("spans.buffer.compression.cpu_time"):
+            compressed = self._zstd_compressor.compress(combined)
+
+        compressed_size = len(compressed)
+
+        compression_ratio = compressed_size / original_size if original_size > 0 else 0
+        metrics.timing("spans.buffer.compression.original_size", original_size)
+        metrics.timing("spans.buffer.compression.compressed_size", compressed_size)
+        metrics.timing("spans.buffer.compression.compression_ratio", compression_ratio)
+
+        return compressed
+
+    def _decompress_batch(self, compressed_data: bytes) -> list[bytes]:
+        # Check for zstd magic header (0xFD2FB528 in little-endian) --
+        # backwards compat with code that did not write compressed payloads.
+        with metrics.timer("spans.buffer.decompression.cpu_time"):
+            if not compressed_data.startswith(b"\x28\xb5\x2f\xfd"):
+                return [compressed_data]
```

Comment:
If you keep the join above, don't you have to split by `\x00` still? I think the assertion in `test_compression_disabled` is wrong, too.